lib/geoengineer/gps/gps.rb

```diff
@@ -0,0 +1,241 @@
+      env = proj[n.environment] ||= {}
+      conf = env[n.configuration] ||= {}
+      nt = conf[n.node_type] ||= {}
+      nt[n.node_name] ||= n.attributes
+    end
+    expanded_hash
+  end
+
+  def loop_projects_hash(projects_hash)
+    # TODO: validate the strucutre before this
+    projects_hash.each_pair do |project, environments|
+      environments.each_pair do |environment, configurations|
+        configurations.each_pair do |configuration, nodes|
+          nodes.each_pair do |node_type, node_names|
+            node_names.each_pair do |node_name, attributes|
+              node_type_class = find_node_class(node_type)
+              node = node_type_class.new(project, environment, configuration, node_name, attributes)
+              yield environments, configurations, nodes, node_names, node
+            end
+          end
+        end
+      end
+    end
+  end
+
+  def expand_meta_nodes(projects_hash)
+    # We dup the original hash because we cannot edit and loop over it at the same time
+    loop_projects_hash(GeoEngineer::GPS.deep_dup(projects_hash)) do |_, _, _, _, node|
+      next unless node.meta?
+      node.validate # ensures that the meta node has expanded and has correct attributes
+
+      # find the hash to edit
+      nodes = projects_hash.dig(node.project, node.environment, node.configuration)
+
+      # node_type => node_name => attrs
+      built_nodes = GeoEngineer::GPS.deep_dup(node.build_nodes) # dup to ensure string keys
+
+      built_nodes.each_pair do |built_node_type, built_node_names|
+        nodes[built_node_type] ||= {}
+        built_node_names.each_pair do |built_node_name, built_attributes|
+          # Error if the meta-node is overwriting an existing node
+          already_built_error = "\"#{node.node_name}\" overwrites node \"#{built_node_name}\""
+          raise MetaNodeError, already_built_error if nodes[built_node_type].key?(built_node_name)
+          # append to the hash
+          nodes[built_node_type][built_node_name] = built_attributes
+        end
+      end
+    end
+
+    projects_hash
+  end
+
+  def build_nodes(projects_hash)
+    all_nodes = []
+    # This is a lot of assumptions
+
+    loop_projects_hash(projects_hash) do |_, _, _, _, node|
+      all_nodes << node
+    end
+
+    all_nodes
+  end
+
+  # returns AND builds the GeoEngineer Project
+  def project(org, name, environment, &block)
+    project_name = "#{org}/#{name}"
+    environment_name = environment.name
+    project_environments = project_environments(project_name)
+
+    raise GPSProjetNotFound, "project not found \"#{project_name}\"" unless project?(project_name)
+
+    project = environment.project(org, name) do
+      environments project_environments
+    end
+
+    # create all resources for projet
+    project_nodes = GeoEngineer::GPS.where(@nodes, "#{project_name}:#{environment_name}:*:*:*")
+    project_nodes.each do |n|
+      n.all_nodes = @nodes
+      n.create_resources(project) unless n.meta?
```

Comment:
a meta node could still override the `create_resources` method I want to be doubly sure that a meta node does not create resources. This makes it a hard constraint rather than a suggestion